oap-server/server-library/library-elasticsearch-client/src/main/java/org/apache/skywalking/library/elasticsearch/bulk/BulkProcessor.java

```diff
@@ -129,57 +135,76 @@ public void flush() {
 
         final List<Holder> batch = new ArrayList<>(requests.size());
         requests.drainTo(batch);
-
-        final CompletableFuture<Void> flush = doFlush(batch);
-        flush.whenComplete((ignored1, ignored2) -> semaphore.release());
-        flush.join();
-
+        final List<CompletableFuture<Void>> futures = doFlush(batch);
+        futures.stream().map(future -> future.join());
+        semaphore.release();
         lastFlushTS = System.currentTimeMillis();
     }
 
-    private CompletableFuture<Void> doFlush(final List<Holder> batch) {
+    private List<CompletableFuture<Void>> doFlush(final List<Holder> batch) {
         log.debug("Executing bulk with {} requests", batch.size());
-
         if (batch.isEmpty()) {
-            return CompletableFuture.completedFuture(null);
+            return Collections.emptyList();
         }
-
-        final CompletableFuture<Void> future = es.get().version().thenCompose(v -> {
-            try {
-                final RequestFactory rf = v.requestFactory();
-                final List<byte[]> bs = new ArrayList<>();
-                for (final Holder holder : batch) {
-                    bs.add(v.codec().encode(holder.request));
-                    bs.add("\n".getBytes());
+        try {
+            int bufferOfBytes = 0;
+            Codec codec = es.get().version().get().codec();
+            final List<byte[]> bs = new ArrayList<>();
+            List<CompletableFuture<Void>> futures = new ArrayList<>();
+            List<ByteBuf> byteBufList = new ArrayList<>();
+            for (final Holder holder : batch) {
+                byte[] bytes = codec.encode(holder.request);
+                bs.add(bytes);
+                bs.add("\n".getBytes());
+                bufferOfBytes += bytes.length;
```

Comment:
You need to also add the `\n`

```suggestion
                bufferOfBytes += bytes.length + 1;
```